pkg/actors/internal/placement.go

```diff
@@ -0,0 +1,374 @@
+	// maintain the status of member by placement.
+	p.shutdownConnLoop.Add(1)
+	go func() {
+		defer p.shutdownConnLoop.Done()
+		for !p.shutdown {
+			// Wait until stream is reconnected.
+			if !p.streamConnAlive || p.clientStream == nil {
+				<-p.streamConnectedCh
+			}
+			if p.shutdown {
+				break
+			}
+
+			// appHealthFn is the health status of actor service application. This allows placement to update
+			// memberlist and hashing table quickly.
+			if !p.appHealthFn() {
+				// app is unresponsive, close the stream and disconnect from the placement service.
+				// Then Placement will remove this host from the member list.
+				log.Debug("disconnecting from placement service by the unhealthy app.")
+
+				p.closeStream()
+				continue
+			}
+
+			host := v1pb.Host{
+				Name:     p.runtimeHostName,
+				Entities: p.actorTypes,
+				Id:       p.appID,
+				Load:     1, // Not used yet
+				// Port is redundant because Name should include port number
+			}
+
+			err := p.clientStream.Send(&host)
+			if err != nil {
+				diag.DefaultMonitoring.ActorStatusReportFailed("send", "status")
+				log.Debugf("failed to report status to placement service : %v", err)
+			}
+
+			time.Sleep(statusReportHeartbeatInterval)
+		}
+	}()
+}
+
+// Stop shuts down server stream gracefully.
+func (p *ActorPlacement) Stop() {
+	p.shutdown = true
+	close(p.streamConnectedCh)
+	p.closeStream()
+	p.shutdownConnLoop.Wait()
+}
+
+func (p *ActorPlacement) closeStream() {
+	if p.clientStream != nil {
+		p.clientStream.CloseSend()
+	}
+
+	if p.clientConn != nil {
+		p.clientConn.Close()
+	}
+}
+
+func (p *ActorPlacement) establishStreamConn() (v1pb.Placement_ReportDaprStatusClient, *grpc.ClientConn) {
+	for !p.shutdown {
+		serverAddr := p.serverAddr[p.serverIndex]
+
+		// Stop reconnecting to placement until app is healthy.
+		if !p.appHealthFn() {
+			time.Sleep(placementReconnectInterval)
+			continue
+		}
+
+		log.Infof("try to connect to placement service: %s", serverAddr)
+
+		opts, err := dapr_credentials.GetClientOptions(p.clientCert, security.TLSServerName)
+		if err != nil {
+			log.Errorf("failed to establish TLS credentials for actor placement service: %s", err)
+			return nil, nil
+		}
+
+		if diag.DefaultGRPCMonitoring.IsEnabled() {
+			opts = append(
+				opts,
+				grpc.WithUnaryInterceptor(diag.DefaultGRPCMonitoring.UnaryClientInterceptor()))
+		}
+
+		conn, err := grpc.Dial(serverAddr, opts...)
+	NEXT_SERVER:
+		if err != nil {
+			log.Debugf("error connecting to placement service: %v", err)
+			conn.Close()
+			p.serverIndex = (p.serverIndex + 1) % len(p.serverAddr)
+			time.Sleep(placementReconnectInterval)
+			continue
+		}
+
+		client := v1pb.NewPlacementClient(conn)
+		stream, err := client.ReportDaprStatus(context.Background())
+		if err != nil {
+			goto NEXT_SERVER
+		}
+
+		log.Infof("established connection to placement service at %q", serverAddr)
+		return stream, conn
```

Comment:
```suggestion
		for {
			if err != nil {
				log.Debugf("error connecting to placement service: %v", err)
				conn.Close()
				p.serverIndex = (p.serverIndex + 1) % len(p.serverAddr)
				time.Sleep(placementReconnectInterval)
				break
			}

			client := v1pb.NewPlacementClient(conn)
			stream, err := client.ReportDaprStatus(context.Background())
			if err != nil {
				continue
			}

			log.Infof("established connection to placement service at %q", serverAddr)
			return stream, conn
		}
```
I am a bit nervous about the goto... would a loop like this work?

Reply:
I avoid to use goto but in this case I do not like to use nested for loop. maybe there is better way to replace it. let me think.